devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java

```diff
@@ -0,0 +1,285 @@
+        return modelName.equals(QuarkusModel.class.getName());
+    }
+
+    @Override
+    public Class getParameterType() {
+        return ModelParameter.class;
+    }
+
+    @Override
+    public Object buildAll(String modelName, Project project) {
+        final ModelParameterImpl modelParameter = new ModelParameterImpl();
+        modelParameter.setMode(LaunchMode.DEVELOPMENT.toString());
+        return buildAll(modelName, modelParameter, project);
+    }
+
+    @Override
+    public Object buildAll(String modelName, Object parameter, Project project) {
+        LaunchMode mode = LaunchMode.valueOf(((ModelParameter) parameter).getMode());
+
+        if (LaunchMode.TEST.equals(mode)) {
+            scannedConfigurations.add(JavaPlugin.TEST_RUNTIME_CLASSPATH_CONFIGURATION_NAME);
+        } else {
+            scannedConfigurations.add(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME);
+        }
+
+        if (LaunchMode.DEVELOPMENT.equals(mode)) {
+            scannedConfigurations.add(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME);
+        }
+
+        final Collection<org.gradle.api.artifacts.Dependency> directExtensionDependencies = getDirectExtension(project);
+
+        final Set<Dependency> appDependencies = new HashSet<>();
+        for (String configurationName : scannedConfigurations) {
+            final ResolvedConfiguration configuration = project.getConfigurations().getByName(configurationName)
+                    .getResolvedConfiguration();
+            appDependencies.addAll(collectDependencies(configuration, configurationName, mode, project));
+            directExtensionDependencies
+                    .addAll(getDirectExtensionDependencies(configuration.getFirstLevelModuleDependencies(), appDependencies));
+        }
+
+        final Set<Dependency> extensionDependencies = collectExtensionDependencies(project, directExtensionDependencies);
+
+        ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
+                project.getVersion().toString());
+
+        return new QuarkusModelImpl(getWorkspace(project),
+                getAdditionalWorkspaces(project.getRootProject(), project.getName()),
+                appDependencies,
+                extensionDependencies);
+    }
+
+    private Workspace getWorkspace(Project project) {
+        ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
+                project.getVersion().toString());
+        final SourceSet mainSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME);
+
+        return new WorkspaceImpl(appArtifactCoords, project.getProjectDir().getAbsoluteFile(),
+                project.getBuildDir().getAbsoluteFile(), getSourceSourceSet(mainSourceSet), convert(mainSourceSet));
+    }
+
+    private Set<Workspace> getAdditionalWorkspaces(Project project, String mainProjectName) {
```

Comment:
There is only one workspace that may include multiple projects.

Reply:
Yes, I will introduce a `project` or `module` type to only keep one `worksapce`